modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  Real angle[n_nodes];
+  unsigned int node_id[3];
+  const Real PI = 3.14159265358979323846264338327950288419716939937510582;
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+    double vertices[3][3];
+
+    for (unsigned int i = 0; i < 3; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < 3; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (abs(angle[i] - PI * 2) > PI * 2 * 0.001)
+    {
+      _boundary_map[i] = new CutPoint;
+      _boundary_map.at(i)->id = i;
+      Node * this_node = _cut_mesh->node_ptr(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < 3; ++j)
+        _boundary_map.at(i)->coord[j] = this_point(j);
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::findBoundaryEdges()
+{
+  _boundary_edges.clear();
+
+  std::vector<unsigned int> special_boundary_elem_id;
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+  unsigned int counter = -1;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    counter++;
+    Elem * cut_elem = *elem_it;
+    unsigned int node_id[3];
+    bool is_node_on_boundary[3];
+
+    for (unsigned int i = 0; i < 3; ++i)
+    {
+      node_id[i] = cut_elem->get_node(i)->id();
+      is_node_on_boundary[i] = (_boundary_map.find(node_id[i]) != _boundary_map.end());
+    }
+
+    if (is_node_on_boundary[0] && is_node_on_boundary[1] && is_node_on_boundary[2])
+    {
+      // store special boundary elements
+      special_boundary_elem_id.push_back(counter);
+    }
+    else
+    {
+      // for other elements, store boundary edges
+      for (unsigned int i = 0; i < 3; ++i)
+      {
+        if (is_node_on_boundary[i] && is_node_on_boundary[(i + 1 <= 2) ? i + 1 : 0])
+        {
+          unsigned int node1 = node_id[i];
+          unsigned int node2 = node_id[(i + 1 <= 2) ? i + 1 : 0];
+          if (node1 > node2)
+            std::swap(node1, node2);
+          _boundary_edges.push_back(new CutEdge);
```

Comment:
You are allocating memory here but I don't see where it's being released! You should not be using raw pointers.